baby-bear/src/aarch64_neon.rs

```diff
@@ -0,0 +1,1071 @@
+use core::arch::aarch64::{self, uint32x4_t};
+use core::iter::{Product, Sum};
+use core::mem::transmute;
+use core::ops::{Add, AddAssign, Div, Mul, MulAssign, Neg, Sub, SubAssign};
+
+use p3_field::{AbstractField, AbstractionOf, Field, PackedField};
+
+use crate::BabyBear;
+
+const WIDTH: usize = 4;
+const P: uint32x4_t = unsafe { transmute::<[u32; WIDTH], _>([0x78000001; WIDTH]) };
+const MU: uint32x4_t = unsafe { transmute::<[u32; WIDTH], _>([0x08000001; WIDTH]) };
+const TOP_BIT: uint32x4_t = unsafe { transmute::<[u32; WIDTH], _>([0x80000000; WIDTH]) };
+
+/// Vectorized NEON implementation of `BabyBear` arithmetic.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)] // This needed to make `transmute`s safe.
+pub struct PackedBabyBearNeon(pub [BabyBear; WIDTH]);
+
+impl PackedBabyBearNeon {
+    #[inline]
+    #[must_use]
+    /// Get an arch-specific vector representing the packed values.
+    fn to_vector(self) -> uint32x4_t {
+        unsafe {
+            // Safety: `BabyBear` is `repr(transparent)` so it can be transmuted to `u32`. It
+            // follows that `[BabyBear; WIDTH]` can be transmuted to `[u32; WIDTH]`, which can be
+            // transmuted to `uint32x4_t`, since arrays are guaranteed to be contiguous in memory.
+            // Finally `PackedBabyBearNeon` is `repr(transparent)` so it can be transmuted to
+            // `[BabyBear; WIDTH]`.
+            transmute(self)
+        }
+    }
+
+    #[inline]
+    #[must_use]
+    /// Make a packed field vector from an arch-specific vector.
+    ///
+    /// SAFETY: The caller must ensure that each element of `vector` represents a valid `BabyBear`.
+    /// In particular, each element of vector must be in `0..P` (canonical form).
+    unsafe fn from_vector(vector: uint32x4_t) -> Self {
+        // Safety: It is up to the user to ensure that elements of `vector` represent valid
+        // `BabyBear` values. We must only reason about memory representations. `uint32x4_t` can be
+        // transmuted to `[u32; WIDTH]` (since arrays elements are contiguous in memory), which can
+        // be transmuted to `[BabyBear; WIDTH]` (since `BabyBear` is `repr(transparent)`), which in
+        // turn can be transmuted to `PackedBabyBearNeon` (since `PackedBabyBearNeon` is also
+        // `repr(transparent)`).
+        transmute(vector)
+    }
+
+    /// Copy `value` to all positions in a packed vector. This is the same as
+    /// `From<BabyBear>::from`, but `const`.
+    #[inline]
+    #[must_use]
+    const fn broadcast(value: BabyBear) -> Self {
+        Self([value; WIDTH])
+    }
+}
+
+impl Add for PackedBabyBearNeon {
+    type Output = Self;
+    #[inline]
+    #[must_use]
```

Comment:
nit: don't think these have an effect in this context, the docs say

> When used on a function in a trait implementation, the attribute does nothing.

Reply:
oh, oops!! i'll remove them